crates/ledger/src/store.rs

```diff
@@ -17,23 +17,76 @@ pub mod columns;
 use super::kernel::{Epoch, Point, PoolId, TransactionInput, TransactionOutput};
 pub use crate::rewards::RewardsSummary;
 use columns::*;
-use std::{borrow::BorrowMut, iter};
+use std::{borrow::BorrowMut, io, iter};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub enum OpenErrorKind {
+    #[error(transparent)]
+    IO(#[from] io::Error),
+    #[error("no ledger stable snapshot found; at least one is expected")]
+    NoStableSnapshot,
+}
+
+#[derive(Error, Debug)]
+pub enum StoreError {
+    #[error(transparent)]
+    Internal(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),
+    #[error("error sending work unit through output port")]
+    Send,
+    #[error("error opening the store")]
+    Open(#[source] OpenErrorKind),
+}
 
 // Store
 // ----------------------------------------------------------------------------
 
-pub trait Store: Send + Sync {
-    type Error: std::error::Error + Send + Sync + 'static;
+pub trait HistoricalStores: Send + Sync {
+    fn for_epoch(&self, epoch: Epoch) -> Result<impl ViewStore, StoreError>;
+}
+
+pub trait ViewStore {
+    /// The most recent snapshot. Note that we never starts from genesis; so there's always a
+    /// snapshot available.
+    fn most_recent_snapshot(&self) -> Epoch;
+
+    /// Get current values of the treasury and reserves accounts.
+    fn with_pots<A>(
+        &self,
+        with: impl FnMut(Box<dyn BorrowMut<pots::Row> + '_>) -> A,
+    ) -> Result<A, StoreError>;
+
+    /// Provide an access to iterate over pools, in a way that enforces:
+    ///
+    /// 1. That mutations will be persisted on-disk
+    ///
+    /// 2. That all operations are consistent and atomic (the iteration occurs on a snapshot, and
+    ///    the mutation apply to the iterated items)
+    fn with_pools(&self, with: impl FnMut(pools::Iter<'_, '_>)) -> Result<(), StoreError>;
+
+    /// Provide an access to iterate over accounts, similar to 'with_pools'.
+    fn with_accounts(&self, with: impl FnMut(accounts::Iter<'_, '_>)) -> Result<(), StoreError>;
 
+    /// Provide an iterator over slot leaders, similar to 'with_pools'. Note that slot leaders are
+    /// stored as a bounded FIFO, so it only make sense to use this function at the end of an epoch
+    /// (or at the beginning, before any block is applied, depending on your perspective).
+    fn with_block_issuers(&self, with: impl FnMut(slots::Iter<'_, '_>)) -> Result<(), StoreError>;
+
+    /// Provide an access to iterate over utxo, similar to 'with_pools'.
+    fn with_utxo(&self, with: impl FnMut(utxo::Iter<'_, '_>)) -> Result<(), StoreError>;
```

Comment:
We could potentially be (i.e. in a subsequent iteration) be smarter here since we won't need to ever _modify_ items in a `ViewStore`. So the with-pattern here with the `BorrowMut` proxies is a bit overkill. Nor do we necessarily needs exactly the same information about those db items (e.g. we may be fine with just the Ada value of UTxO, although now I am perhaps over-fitting to the rewards calculation...).

Reply:
(reviewed further down the PR): I see; I thought `ViewStore` represented the immutable snapshots, but no, it's capturing some "query layer" on top of the store; 

Then, I am not sure to fully understand the boundary between `Store` and `ViewStore` given that there are still accessor methods in `Store` (e.g. `pool`, `resolve_inputs` ...) ?

Reply:
Right, currently I just migrated what was strictly needed for the rewards calculation. How about we add the other read-only fns to the trait?

Reply:
That would make the frontier clearer to me 😅 ? At first, I thought the split was Read-only / Read-write, but no; then I thought it's database-creation vs accesses, but no. 

So perhaps it's about having a `ReadWriteStore` and `ReadOnlyStore`?